### PR TITLE
feat(person-on-events): Person on events trends

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- scoped down the work to just account for trends
- example test case
- Adds a personpropertiesmode
- Adds a setting for teams to whitelist

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: ee/clickhouse/queries/trends/breakdown.py [hunk 1]
```diff
@@ -358,3 +374,43 @@ def _person_join_condition(self) -> Tuple[str, Dict]:
             return event_join, {}
         else:
             return "", {}
+
+    def _groups_join_conditino(self) -> Tuple[str, Dict]:
```
Comment (neilkakkar) [inline ee/clickhouse/queries/trends/breakdown.py:378]:
```suggestion
    def _groups_join_condition(self) -> Tuple[str, Dict]:
```

File: ee/clickhouse/queries/trends/breakdown.py [hunk 2]
```diff
@@ -358,3 +374,43 @@ def _person_join_condition(self) -> Tuple[str, Dict]:
...
+
+class ClickhouseTrendsBreakdown_PersonsOnEvents(ClickhouseTrendsBreakdown):
+    @cached_property
+    def _person_properties_mode(self) -> PersonPropertiesMode:
+        return PersonPropertiesMode.DIRECT_ON_EVENTS
+
+    def _person_join_condition(self) -> Tuple[str, Dict]:
+        return "", {}
+
+    def _groups_join_conditino(self) -> Tuple[str, Dict]:
```
Comment (neilkakkar) [inline ee/clickhouse/queries/trends/breakdown.py:390]:
```suggestion
    def _groups_join_condition(self) -> Tuple[str, Dict]:
```

File: ee/clickhouse/sql/trends/breakdown.py [hunk 3]
```diff
@@ -106,10 +106,8 @@
         SELECT toStartOfDay(toDateTime(timestamp), %(timezone)s) as timestamp FROM events e WHERE team_id = %(team_id)s {parsed_date_from_prev_range} {parsed_date_to} GROUP BY timestamp
     ) d
     CROSS JOIN (
-        SELECT toStartOfDay(toDateTime(timestamp), %(timezone)s) as timestamp, pdi.person_id AS person_id, {breakdown_value} as breakdown_value
+        SELECT toStartOfDay(toDateTime(timestamp), %(timezone)s) as timestamp, {person_id_alias}.person_id AS person_id, {breakdown_value} as breakdown_value
         FROM events e
-        INNER JOIN ({GET_TEAM_PERSON_DISTINCT_IDS}) as pdi
```
Comment (neilkakkar) [inline ee/clickhouse/sql/trends/breakdown.py:111]: this was a nasty bug. Always having these here meant we'll have two pdi joins when we also want to join persons

https://sentry.io/organizations/posthog2/issues/3154668145/?query=is%3Aunresolved

File: ee/clickhouse/queries/trends/person.py [hunk 4]
```diff
@@ -95,10 +95,11 @@ def actor_query(self, limit_actors: Optional[bool] = True) -> Tuple[str, Dict]:
             filter=self._filter,
             team=self._team,
             entity=self.entity,
-            should_join_distinct_ids=not self.is_aggregating_by_groups,
-            should_join_persons=not self.is_aggregating_by_groups,
```
Comment (neilkakkar) [inline ee/clickhouse/queries/trends/person.py:99]: Person join only needs to happen if we have person properties to handle, not every time. (Notice snapshot tests being updated, where the person join was unnecessary)

File: posthog/queries/event_query.py [hunk 5]
```diff
@@ -59,7 +60,7 @@ def __init__(
         self._should_join_distinct_ids = should_join_distinct_ids
         self._should_join_persons = should_join_persons
         self._extra_fields = extra_fields
-        self._extra_person_fields = extra_person_fields
```
Comment (neilkakkar) [inline posthog/queries/event_query.py:62]: annoying diff: extra_person_fields are set in a line above

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_trends.ambr [hunk 6]
```diff
@@ -186,7 +186,7 @@
      GROUP BY entrance_period_start) data
   RIGHT OUTER JOIN
     (SELECT toStartOfDay(toDateTime('2021-04-30 07:00:00') + toIntervalDay(number), 'US/Pacific') AS entrance_period_start
-     FROM numbers(dateDiff('day', toDateTime('2021-04-30 07:00:00'), toDateTime('2021-05-08 06:59:59')) + 1) AS period_offsets) fill USING (entrance_period_start)
+     FROM numbers(dateDiff('day', toDateTime('2021-04-30 07:00:00'), toDateTime('2021-05-07 07:00:00')) + 1) AS period_offsets) fill USING (entrance_period_start)
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_trends.ambr:189]: hmm, why are these changes happening in my PR? 🤔 

CH version change issue?

Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_trends.ambr:189]: Yes: https://github.com/PostHog/posthog/pull/9885

Comment (EDsCODE) [reply ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_trends.ambr:189]: 👍

File: ee/clickhouse/queries/breakdown_props.py [hunk 7]
```diff
@@ -109,19 +130,29 @@ def _to_value_expression(
...
     breakdown_group_type_index: Optional[GroupTypeIndex],
+    direct_on_events: bool = False,
 ) -> str:
     if breakdown_type == "person":
-        return get_single_or_multi_property_string_expr(breakdown, table="person", query_alias="value")
+        return get_single_or_multi_property_string_expr(
+            breakdown,
+            table="events" if direct_on_events else "person",
+            query_alias="value",
+            column="person_properties" if direct_on_events else "person_props",
```
Comment (EDsCODE) [inline ee/clickhouse/queries/breakdown_props.py:140]: Hmm I wonder if there's a better way to do this column determination. Now, we have to spam some sort of conditional every time we call `get_single_or_multi_property_string_expr`

File: ee/clickhouse/models/event.py [hunk 8]
```diff
@@ -96,22 +97,49 @@ def bulk_create_events(events: List[Dict[str, Any]], person_mapping: Optional[Di
...
+        else:
+            try:
+                person = Person.objects.get(
+                    persondistinctid__distinct_id=event["distinct_id"], persondistinctid__team_id=team_id
+                )
+                person_properties = person.properties
+                person_id = person.uuid
+            except Person.DoesNotExist:
+                person_properties = {}
+                person_id = cast(uuid.UUID, "00000000-0000-0000-0000-000000000000")
```
Comment (EDsCODE) [inline ee/clickhouse/models/event.py:114]: something interesting I found in funnels PR that may need to be accounted for is that null personids could end up being counted in unexpected ways. For example, on trends, if we're looking at unique users (DISTINCT events.person_id), all anonymous users will get collapsed into the same user

File: ee/clickhouse/queries/test/test_trends.py [hunk 9]
```diff
@@ -134,6 +147,11 @@ def test_breakdown_by_group_props(self):
         )
         response = ClickhouseTrends().run(filter, self.team,)
 
+        with override_instance_config("ENABLE_ACTOR_ON_EVENTS_TEAMS", f"{self.team.pk}"):
+            new_response = ClickhouseTrends().run(filter, self.team)
+
+        self.assertEqual(response, new_response)
```
Comment (EDsCODE) [inline ee/clickhouse/queries/test/test_trends.py:153]: Do we still want this now that we have new testing pattern?

File: ee/clickhouse/models/event.py (hunk 8 above)
Comment (neilkakkar) [reply ee/clickhouse/models/event.py:114]: Do anonymous users not have personIDs? 🤔

Comment (neilkakkar) [reply ee/clickhouse/models/event.py:114]: Seems like we're good here: https://posthog.slack.com/archives/C03E27SM90S/p1653059142798669

Need to fix this though - no case for person_id being `0000...`.

File: ee/clickhouse/queries/test/test_trends.py (hunk 9 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/test/test_trends.py:153]: Not really. Will get rid of it

File: ee/clickhouse/queries/breakdown_props.py (hunk 7 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/breakdown_props.py:140]: I thought being explicit here is better than magic defaults or overriding based on a parameter 😅 . made things clearer

Comment (neilkakkar) [reply ee/clickhouse/queries/breakdown_props.py:140]: .. and keeps this function dumb, which I also like

Comment (neilkakkar) [reply ee/clickhouse/queries/breakdown_props.py:140]: Follow up TODO: `all` cohort doesn't seem to have tests / might blow up here. Also need to check if we still support this on the frontend

Comment (EDsCODE) [reply ee/clickhouse/queries/breakdown_props.py:140]: fair let's roll with it then

Comment (neilkakkar) [reply ee/clickhouse/queries/breakdown_props.py:140]: image

**Comment by EDsCODE** [comment]
@neilkakkar  This should be good to go!

File: posthog/queries/test/test_trends.py [hunk 10]
```diff
@@ -1849,6 +1814,20 @@ def test_person_property_filtering(self):
                     ),
                     self.team,
                 )
+
+            with freeze_time("2020-01-04"), override_instance_config("ENABLE_ACTOR_ON_EVENTS_TEAMS", f"{self.team.pk}"):
```
Comment (EDsCODE) [inline posthog/queries/test/test_trends.py:1818]: One more

Comment (neilkakkar) [reply posthog/queries/test/test_trends.py:1818]: will remove in a follow up 🙈

**Pull request merged**